fontra2fontir/src/toir.rs

```diff
@@ -0,0 +1,265 @@
+            .iter()
+            .map(|c| to_ir_path(fontra_glyph.name.clone(), c))
+            .collect::<Result<_, _>>()?;
+        instances.insert(
+            location.clone(),
+            GlyphInstance {
+                width: layer.glyph.x_advance,
+                contours,
+                ..Default::default()
+            },
+        );
+    }
+
+    Glyph::new(fontra_glyph.name.clone(), true, codepoints, instances)
+}
+
+#[allow(dead_code)] // TEMPORARY
+fn add_to_path<'a>(
+    glyph_name: GlyphName,
+    path_builder: &'a mut GlyphPathBuilder,
+    points: impl Iterator<Item = &'a FontraPoint>,
+) -> Result<(), WorkError> {
+    // Walk through the remaining points, accumulating off-curve points until we see an on-curve
+    // https://github.com/googlefonts/glyphsLib/blob/24b4d340e4c82948ba121dcfe563c1450a8e69c9/Lib/glyphsLib/pens.py#L92
+    for point in points {
+        let point_type = point
+            .point_type()
+            .map_err(|e| WorkError::InvalidSourceGlyph {
+                glyph_name: glyph_name.clone(),
+                message: format!("No point type for {point:?}: {e}"),
+            })?;
+        // Smooth is only relevant to editors so ignore here
+        match point_type {
+            PointType::OnCurve | PointType::OnCurveSmooth => path_builder
+                .curve_to((point.x, point.y))
+                .map_err(WorkError::PathConversionError)?,
+            PointType::OffCurveQuad | PointType::OffCurveCubic => path_builder
+                .offcurve((point.x, point.y))
+                .map_err(WorkError::PathConversionError)?,
+        }
+    }
+    Ok(())
+}
+
+fn to_ir_path(glyph_name: GlyphName, contour: &FontraContour) -> Result<BezPath, WorkError> {
+    // Based on glyphs2fontir/src/toir.rs to_ir_path
+    // TODO: so similar a trait to to let things be added to GlyphPathBuilder would be nice
+    if contour.points.is_empty() {
+        return Ok(BezPath::new());
+    }
+
+    let mut path_builder = GlyphPathBuilder::new(glyph_name.clone(), contour.points.len());
+
+    if !contour.is_closed {
+        let first = contour.points.first().unwrap();
+        let first_type = first
+            .point_type()
+            .map_err(|e| WorkError::InvalidSourceGlyph {
+                glyph_name: glyph_name.clone(),
+                message: format!("No point type for {first:?}: {e}"),
+            })?;
+        if first_type.is_off_curve() {
+            return Err(WorkError::InvalidSourceGlyph {
+                glyph_name: glyph_name.clone(),
+                message: String::from("Open path starts with off-curve points"),
+            });
+        }
+        path_builder.move_to((first.x, first.y))?;
+        add_to_path(
+            glyph_name.clone(),
+            &mut path_builder,
+            contour.points[1..].iter(),
+        )?;
+    } else {
+        // I think Fontra does the start at the end bit
+        // Rotate right by 1 by way of chaining iterators
```

Comment:
where did you get that impression? I think it just follows ufo, where the open path is signalled by the first point having a "move" type, whereas closed path can start with anything (doesn't really start or ends, it's a cyclical list of points) 

See how a fontra PackedPath is drawn onto a (UFO) PointPen:

https://github.com/googlefonts/fontra/blob/a4edd06837118e583804fd963c22ed806a315b04/src/fontra/core/path.py#L137-L179

and this is the pen that constructs a fontra PackedPath from a set of PointPen commands:

https://github.com/googlefonts/fontra/blob/a4edd06837118e583804fd963c22ed806a315b04/src/fontra/core/path.py#L301-L382

I don't see any rotation occurring there, but maybe I'm wrong and you're right. /cc @justvanrossum

Reply:
I have no idea what this code is doing, but: Fontra contours have an explicit `isClosed` attribute.

Reply:
Fontra does not rotate. At least as long as we stick to the a-contour-is-a-list-of-points view.

Reply:
thanks for confirming. But at least for open paths it's invalid to have leading/trailing off curves, right?

Reply:
Fixing, ty for confirming

Reply:
> But at least for open paths it's invalid to have leading/trailing off curves, right?

Sorry, no, leading/trailing off-curve points are perfectly legal in Fontra... For rendering you can just strip them, though. This is only important for a work-in-progress contour.